File: server/src/main/java/com/yawa/validators/Role.java
package com.yawa.validators;

import javax.validation.Constraint;
import javax.validation.Payload;
import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Target({ FIELD })
@Retention(RUNTIME)
@Constraint(validatedBy = RoleValidator.class)
@Documented
public @interface Role {

    String message() default RoleValidator.VALIDATION_ERROR_MESSAGE;

    Class<?>[] groups() default { };

    Class<? extends Payload>[] payload() default { };

}

File: server/src/main/kotlin/com/yawa/server/validators/Role.kt
package com.yawa.server.validators

import com.yawa.server.models.users.UserRole
import org.apache.commons.lang3.StringUtils
import java.util.regex.Pattern
import javax.validation.Constraint
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext
import javax.validation.Payload
import kotlin.reflect.KClass

@Target(AnnotationTarget.PROPERTY, AnnotationTarget.VALUE_PARAMETER)
@MustBeDocumented
@Constraint(validatedBy = [RoleValidator::class])
annotation class Role(
    val message: String = "{com.yawa.server.validators.RoleValidator.VALIDATION_ERROR_MESSAGE}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)

class RoleValidator : ConstraintValidator<Role, String> {

    val PATTERN : Pattern = Pattern.compile("^${StringUtils.join(UserRole.values(), "|")}$")
    val VALIDATION_ERROR_MESSAGE = "Invalid role: must be one of ${UserRole.values()}"

    override fun isValid(value: String, context: ConstraintValidatorContext): Boolean =
        PATTERN.matcher(value).matches()
}
